File: app/api/routes/leads.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.api.deps import get_app_settings, get_db_session
from app.config import Settings
from app.enums import CompanySizeFit, LeadSourceType, LeadStatus, TradeType
from app.repositories.lead_repository import LeadRepository
from app.schemas.lead import (
    LeadAssignmentRunResult,
    LeadAssignmentSuggestionRead,
    LeadBatchAssignmentRequest,
    LeadBatchAssignmentResponse,
    LeadBatchAssignmentSummary,
    LeadBatchEnrichmentRequest,
    LeadBatchEnrichmentResponse,
    LeadBlockedFilter,
    LeadDetail,
    LeadImportBatchResponse,
    LeadListFilters,
    LeadListResponse,
    LeadOptionsResponse,
    LeadScopeRequest,
    LeadScopeResolveResponse,
    LeadSummary,
    LeadSortBy,
    LeadSortDir,
    LeadUpdateRequest,
)
from app.services.crm import CRMService
from app.services.enrichment import EnrichmentService
from app.services.lead_assignment import LeadAssignmentService
from app.services.lead_scope import LeadScopeResolver, ResolvedLeadScope

router = APIRouter(prefix="/leads", tags=["leads"])


@router.get("", response_model=LeadListResponse)
def list_leads(
    city: str | None = Query(default=None),
    state: str | None = Query(default=None),
    status: LeadStatus | None = Query(default=None),
    has_email: bool | None = Query(default=None),
    has_whatsapp: bool | None = Query(default=None),
    has_instagram: bool | None = Query(default=None),
    category: str | None = Query(default=None),
    score_min: int | None = Query(default=None, ge=0, le=100),
    score_max: int | None = Query(default=None, ge=0, le=100),
    lead_source_type: LeadSourceType | None = Query(default=None),
    do_not_contact: bool | None = Query(default=None),
    sales_region_id: int | None = Query(default=None),
    market_segment_id: int | None = Query(default=None),
    market_subsegment_id: int | None = Query(default=None),
    assigned_sales_rep_id: int | None = Query(default=None),
    has_assignment: bool | None = Query(default=None),
    company_size_fit: CompanySizeFit | None = Query(default=None),
    trade_type: TradeType | None = Query(default=None),
    blocked: LeadBlockedFilter = Query(default="exclude"),
    sort_by: LeadSortBy = Query(default="updated_at"),
    sort_dir: LeadSortDir = Query(default="desc"),
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    db: Session = Depends(get_db_session),
) -> LeadListResponse:
    filters = LeadListFilters(
        city=city,
        state=state,
        status=status,
        has_email=has_email,
        has_whatsapp=has_whatsapp,
        has_instagram=has_instagram,
        category=category,
        score_min=score_min,
        score_max=score_max,
        lead_source_type=lead_source_type,
        do_not_contact=do_not_contact,
        sales_region_id=sales_region_id,
        market_segment_id=market_segment_id,
        market_subsegment_id=market_subsegment_id,
        assigned_sales_rep_id=assigned_sales_rep_id,
        has_assignment=has_assignment,
        company_size_fit=company_size_fit,
        trade_type=trade_type,
        blocked=blocked,
        sort_by=sort_by,
        sort_dir=sort_dir,
        limit=limit,
        offset=offset,
    )
    repository = LeadRepository(db)
    leads, total = repository.list_leads(filters)
    return LeadListResponse(total=total, items=[LeadSummary.model_validate(lead) for lead in leads])


@router.get("/options", response_model=LeadOptionsResponse)
def get_lead_options(db: Session = Depends(get_db_session)) -> LeadOptionsResponse:
    repository = LeadRepository(db)
    assignment_options = repository.list_v2_filter_options()
    return LeadOptionsResponse(
        cities=repository.list_distinct_cities(),
        states=repository.list_distinct_states(),
        statuses=[status.value for status in LeadStatus],
        assigned_reps=assignment_options["assigned_reps"],
        sales_regions=assignment_options["sales_regions"],
        market_segments=assignment_options["market_segments"],
        market_subsegments=assignment_options["market_subsegments"],
        target_fit_values=[value.value for value in CompanySizeFit],
        trade_type_values=[value.value for value in TradeType],
    )


@router.get("/import-batches/latest", response_model=LeadImportBatchResponse)
def get_latest_import_batch(db: Session = Depends(get_db_session)) -> LeadImportBatchResponse:
    repository = LeadRepository(db)
    batch = repository.get_latest_completed_import_batch()
    if batch is None:
        raise HTTPException(status_code=404, detail="No completed import batch with leads was found.")
    lead_ids = repository.list_lead_ids_for_import_batch(batch.id)
    return _import_batch_response(batch, lead_ids)


@router.post("/resolve-scope", response_model=LeadScopeResolveResponse)
def resolve_lead_scope(
    payload: LeadScopeRequest,
    db: Session = Depends(get_db_session),
) -> LeadScopeResolveResponse:
    try:
        resolved = LeadScopeResolver(db).resolve(payload)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return _scope_response(resolved)


@router.post("/batch/enrich", response_model=LeadBatchEnrichmentResponse)
def enrich_lead_batch(
    payload: LeadBatchEnrichmentRequest,
    db: Session = Depends(get_db_session),
    settings: Settings = Depends(get_app_settings),
) -> LeadBatchEnrichmentResponse:
    service = EnrichmentService(db=db, settings=settings)
    try:
        response = service.enrich_lead_ids(
            payload.lead_ids,
            actor="api",
            scope_label="api lead ids",
            continue_on_error=False,
        )
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return response


@router.post("/batch/assign", response_model=LeadBatchAssignmentResponse)
def assign_lead_batch(
    payload: LeadBatchAssignmentRequest,
    db: Session = Depends(get_db_session),
) -> LeadBatchAssignmentResponse:
    try:
        resolved = LeadScopeResolver(db).resolve(payload)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc

    service = LeadAssignmentService(db)
    result = service.apply_batch(
        lead_ids=resolved.lead_ids,
        overwrite=payload.overwrite,
        dry_run=payload.dry_run,
    )
    if not payload.dry_run:
        db.commit()
    return LeadBatchAssignmentResponse(
        processed=result.processed,
        changed=result.changed,
        dry_run=result.dry_run,
        results=[
            LeadAssignmentRunResult(
                lead_id=item.lead_id,
                changed_fields=item.changed_fields,
                suggestion=LeadAssignmentSuggestionRead(
                    sales_region_id=item.suggestion.sales_region_id,
                    market_segment_id=item.suggestion.market_segment_id,
                    market_subsegment_id=item.suggestion.market_subsegment_id,
                    assigned_sales_rep_id=item.suggestion.assigned_sales_rep_id,
                    assignment_rule_id=item.suggestion.assignment_rule_id,
                    explanation=item.suggestion.explanation,
                    metadata=item.suggestion.metadata,
                ),
            )
            for item in result.results
        ],
        summary=LeadBatchAssignmentSummary(
            scope_type=resolved.scope_type,
            scope_label=resolved.scope_label,
            requested=resolved.requested_count,
            processed=result.processed,
            changed=result.changed,
            overwrite=payload.overwrite,
            dry_run=payload.dry_run,
            missing_lead_ids=resolved.missing_lead_ids,
        ),
    )


@router.get("/{lead_id}", response_model=LeadDetail)
def get_lead_detail(lead_id: int, db: Session = Depends(get_db_session)) -> LeadDetail:
    repository = LeadRepository(db)
    lead = repository.get_detail(lead_id)
    if lead is None:
        raise HTTPException(status_code=404, detail=f"Lead {lead_id} not found.")
    return LeadDetail.model_validate(lead)


@router.patch("/{lead_id}", response_model=LeadDetail)
def update_lead(
    lead_id: int,
    payload: LeadUpdateRequest,
    db: Session = Depends(get_db_session),
) -> LeadDetail:
    service = CRMService(db)
    try:
        lead = service.update_lead(lead_id, payload, actor="api")
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return LeadDetail.model_validate(lead)


@router.post("/{lead_id}/enrich", response_model=LeadBatchEnrichmentResponse)
def enrich_single_lead(
    lead_id: int,
    db: Session = Depends(get_db_session),
    settings: Settings = Depends(get_app_settings),
) -> LeadBatchEnrichmentResponse:
    service = EnrichmentService(db=db, settings=settings)
    try:
        response = service.enrich_lead_ids(
            [lead_id],
            actor="api",
            scope_label="api single lead",
            continue_on_error=False,
        )
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return response


def _scope_response(resolved: ResolvedLeadScope) -> LeadScopeResolveResponse:
    import_batch = (
        _import_batch_response(resolved.import_batch, resolved.lead_ids)
        if resolved.import_batch is not None
        else None
    )
    return LeadScopeResolveResponse(
        scope_type=resolved.scope_type,
        scope_label=resolved.scope_label,
        total=len(resolved.lead_ids),
        lead_ids=resolved.lead_ids,
        missing_lead_ids=resolved.missing_lead_ids,
        import_batch=import_batch,
    )


def _import_batch_response(batch, lead_ids: list[int]) -> LeadImportBatchResponse:
    return LeadImportBatchResponse(
        id=batch.id,
        batch_type=batch.batch_type,
        status=batch.status,
        source_provider=batch.source_provider,
        source_query=batch.source_query,
        location_label=batch.location_label,
        record_count=batch.record_count,
        lead_count=len(lead_ids),
        lead_ids=lead_ids,
        started_at=batch.started_at,
        completed_at=batch.completed_at,
        created_at=batch.created_at,
        updated_at=batch.updated_at,
    )

File: web/lib/demo/leads.ts
import type {
  LeadBatchAssignmentResponse,
  LeadBatchCNPJEnrichmentResponse,
  LeadBatchEnrichmentResponse,
  LeadDetail,
  LeadImportBatchResponse,
  LeadListParams,
  LeadListResponse,
  LeadOptionsResponse,
  LeadScopeRequest,
  LeadScopeResolveResponse,
} from "@/lib/api/types";
import {
  demoInitialImportBatch,
  demoMarketSegments,
  demoMarketSubsegments,
  demoSalesRegions,
  demoSalesReps,
} from "@/lib/demo/fixtures";
import { getDemoImportBatches, getDemoLeads, saveDemoLeads } from "@/lib/demo/storage";

export async function listDemoLeads(params: LeadListParams = {}): Promise<LeadListResponse> {
  const filtered = filterDemoLeads(getDemoLeads(), params);
  const sorted = sortDemoLeads(filtered, params);
  const offset = params.offset ?? 0;
  const limit = params.limit ?? sorted.length;
  return {
    total: sorted.length,
    items: sorted.slice(offset, offset + limit),
  };
}

export async function getDemoLeadOptions(): Promise<LeadOptionsResponse> {
  const leads = getDemoLeads();
  return {
    cities: unique(leads.map((lead) => lead.city).filter(Boolean) as string[]),
    states: unique(leads.map((lead) => lead.state).filter(Boolean) as string[]),
    statuses: ["new", "reviewed", "approved", "contacted", "replied", "interested", "closed", "not_interested", "do_not_contact"],
    assigned_reps: demoSalesReps.map(({ id, name }) => ({ id, name })),
    sales_regions: demoSalesRegions,
    market_segments: demoMarketSegments,
    market_subsegments: demoMarketSubsegments,
    target_fit_values: ["ideal_sme", "possible_sme", "large_enterprise", "unknown"],
    trade_type_values: ["varejo", "atacado", "distribuidora", "ecommerce", "industria", "construcao_civil", "unknown"],
  };
}

export async function getDemoLeadDetail(leadId: number): Promise<LeadDetail> {
  const lead = getDemoLeads().find((item) => item.id === leadId);
  if (!lead) {
    throw new Error("Demo lead not found.");
  }
  return lead;
}

export async function getDemoLatestImportBatch(): Promise<LeadImportBatchResponse> {
  return getDemoImportBatches()[0] ?? demoInitialImportBatch;
}

export async function resolveDemoLeadScope(scope: LeadScopeRequest): Promise<LeadScopeResolveResponse> {
  if ("lead_ids" in scope && scope.lead_ids) {
    const existingIds = new Set(getDemoLeads().map((lead) => lead.id));
    const leadIds = scope.lead_ids.filter((leadId) => existingIds.has(leadId));
    return {
      scope_type: "selected",
      scope_label: "Selected demo leads",
      total: leadIds.length,
      lead_ids: leadIds,
      missing_lead_ids: scope.lead_ids.filter((leadId) => !existingIds.has(leadId)),
      import_batch: null,
    };
  }

  if ("latest_import_batch" in scope && scope.latest_import_batch) {
    const latestBatch = await getDemoLatestImportBatch();
    return {
      scope_type: "latest_import_batch",
      scope_label: `Latest import batch #${latestBatch.id}`,
      total: latestBatch.lead_ids.length,
      lead_ids: latestBatch.lead_ids,
      missing_lead_ids: [],
      import_batch: latestBatch,
    };
  }

  const filtered = filterDemoLeads(getDemoLeads(), scope.filters ?? {});
  return {
    scope_type: "filters",
    scope_label: "Current filtered demo list",
    total: filtered.length,
    lead_ids: filtered.map((lead) => lead.id),
    missing_lead_ids: [],
    import_batch: null,
  };
}

export async function enrichDemoLeadBatch(leadIds: number[]): Promise<LeadBatchEnrichmentResponse> {
  const now = new Date().toISOString();
  const leads = getDemoLeads();
  const selected = leads.filter((lead) => leadIds.includes(lead.id));
  const updated = leads.map((lead) => leadIds.includes(lead.id) ? { ...lead, last_enriched_at: now, updated_at: now } : lead);
  saveDemoLeads(updated);
  const emailsFound = selected.filter((lead) => lead.email).length;
  const instagramsFound = selected.filter((lead) => lead.instagram).length;
  const whatsappsFound = selected.filter((lead) => lead.whatsapp).length;

  return {
    processed: selected.length,
    results: selected.map((lead) => ({
      lead_id: lead.id,
      business_name: lead.business_name,
      success: true,
      pages_attempted: lead.website ? 2 : 0,
      pages_fetched: lead.website ? 2 : 0,
      attempted_pages: [],
      fetched_page_urls: lead.website ? [lead.website] : [],
      extracted_contacts: [],
      contacts_added: 0,
      contacts_added_by_type: {},
      fields_updated: ["last_enriched_at"],
      last_enriched_at: now,
      material_profile: {},
      skipped_reason: lead.website ? null : "No public website.",
      no_email_found: !lead.email,
      error_message: null,
    })),
    summary: {
      scope_label: "Demo enrichment",
      requested: leadIds.length,
      processed: selected.length,
      success_count: selected.length,
      contacts_added: 0,
      emails_found: emailsFound,
      instagrams_found: instagramsFound,
      whatsapps_found: whatsappsFound,
      contact_forms_found: selected.filter((lead) => lead.website).length,
      skipped: Math.max(0, leadIds.length - selected.length),
      skipped_no_website: selected.filter((lead) => !lead.website).length,
      errors: 0,
      error_messages: [],
      failed_lead_ids: [],
      pages_attempted: selected.filter((lead) => lead.website).length * 2,
      pages_fetched: selected.filter((lead) => lead.website).length * 2,
    },
  };
}

export async function enrichDemoLeadBatchCnpj(leadIds: number[]): Promise<LeadBatchCNPJEnrichmentResponse> {
  return {
    processed: leadIds.length,
    results: [],
    summary: {
      scope_label: "Demo CNPJ enrichment disabled",
      requested: leadIds.length,
      processed: 0,
      matched_count: 0,
      needs_review_count: 0,
      not_found_count: 0,
      skipped_known_count: 0,
      skipped_review_candidate_count: 0,
      paid_search_recently_attempted_count: 0,
      no_website_count: 0,
      no_cnpj_on_website_count: 0,
      website_timeout_count: 0,
      website_unreachable_count: 0,
      validation_failed_count: 0,
      low_confidence_count: 0,
      company_search_matched_count: 0,
      company_search_needs_review_count: 0,
      company_search_no_candidates_count: 0,
      company_search_zero_candidates_count: 0,
      company_search_low_confidence_count: 0,
      company_search_not_configured_count: 0,
      company_search_pending_retry_count: 0,
      company_search_rate_limited_count: 0,
      company_search_provider_error_count: 0,
      company_search_consulted_now_count: 0,
      paid_calls_made: 0,
      paid_calls_skipped_duplicate: 0,
      paid_calls_skipped_recent: 0,
      provider_rate_limited_count: 0,
      provider_error_count: 0,
      error_count: 0,
      errors: [],
    },
  };
}

export async function assignDemoLeadBatch(scope: LeadScopeRequest): Promise<LeadBatchAssignmentResponse> {
  const resolved = await resolveDemoLeadScope(scope);
  const now = new Date().toISOString();
  const leads = getDemoLeads();
  const updated = leads.map((lead) => {
    if (!resolved.lead_ids.includes(lead.id) || lead.assigned_sales_rep_id) {
      return lead;
    }
    return {
      ...lead,
      assigned_sales_rep_id: 1,
      assigned_sales_rep: demoSalesReps[0],
      sales_region_id: lead.sales_region_id ?? 1,
      sales_region: lead.sales_region ?? demoSalesRegions[0],
      assigned_at: now,
      updated_at: now,
    };
  });
  saveDemoLeads(updated);
  const changed = updated.filter((lead) => resolved.lead_ids.includes(lead.id) && lead.assigned_sales_rep_id === 1).length;
  return {
    processed: resolved.lead_ids.length,
    changed,
    dry_run: false,
    results: resolved.lead_ids.map((leadId) => ({
      lead_id: leadId,
      changed_fields: ["assigned_sales_rep_id"],
      suggestion: {
        sales_region_id: 1,
        market_segment_id: null,
        market_subsegment_id: null,
        assigned_sales_rep_id: 1,
        assignment_rule_id: null,
        explanation: "Demo assignment suggestion.",
        metadata: {},
      },
    })),
    summary: {
      scope_type: resolved.scope_type,
      scope_label: resolved.scope_label,
      requested: resolved.total,
      processed: resolved.total,
      changed,
      overwrite: false,
      dry_run: false,
      missing_lead_ids: resolved.missing_lead_ids,
    },
  };
}

export async function approveDemoLeadCnpjCandidate(leadId: number): Promise<LeadDetail> {
  return getDemoLeadDetail(leadId);
}

export async function rejectDemoLeadCnpjCandidate(leadId: number): Promise<LeadDetail> {
  return getDemoLeadDetail(leadId);
}

export function filterDemoLeads(leads: LeadDetail[], params: LeadListParams) {
  let filtered = leads;
  if (params.blocked === "exclude" || !params.blocked) {
    filtered = filtered.filter((lead) => !lead.is_blocked);
  } else if (params.blocked === "only") {
    filtered = filtered.filter((lead) => lead.is_blocked);
  }
  if (params.search) {
    const search = params.search.toLocaleLowerCase("pt-BR");
    filtered = filtered.filter((lead) => [
      lead.business_name,
      lead.city,
      lead.state,
      lead.category,
      lead.email,
      lead.phone,
      lead.whatsapp,
      lead.instagram,
      lead.assigned_sales_rep?.name,
      lead.market_segment?.name,
      lead.blocked_reason,
    ].filter(Boolean).some((value) => String(value).toLocaleLowerCase("pt-BR").includes(search)));
  }
  if (params.city) filtered = filtered.filter((lead) => lead.city === params.city);
  if (params.state) filtered = filtered.filter((lead) => lead.state === params.state);
  if (params.status) filtered = filtered.filter((lead) => lead.status === params.status);
  if (params.assigned_sales_rep_id) filtered = filtered.filter((lead) => lead.assigned_sales_rep_id === params.assigned_sales_rep_id);
  if (params.sales_region_id) filtered = filtered.filter((lead) => lead.sales_region_id === params.sales_region_id);
  if (params.market_segment_id) filtered = filtered.filter((lead) => lead.market_segment_id === params.market_segment_id);
  if (params.market_subsegment_id) filtered = filtered.filter((lead) => lead.market_subsegment_id === params.market_subsegment_id);
  if (params.company_size_fit) filtered = filtered.filter((lead) => lead.company_size_fit === params.company_size_fit);
  if (params.trade_type) filtered = filtered.filter((lead) => lead.trade_type === params.trade_type);
  if (params.has_assignment !== undefined) filtered = filtered.filter((lead) => Boolean(lead.assigned_sales_rep_id) === params.has_assignment);
  if (params.has_email !== undefined) filtered = filtered.filter((lead) => Boolean(lead.email) === params.has_email);
  if (params.has_whatsapp !== undefined) filtered = filtered.filter((lead) => Boolean(lead.whatsapp) === params.has_whatsapp);
  if (params.has_instagram !== undefined) filtered = filtered.filter((lead) => Boolean(lead.instagram) === params.has_instagram);
  if (params.import_batch_id) {
    const batch = getDemoImportBatches().find((item) => item.id === params.import_batch_id);
    const ids = new Set(batch?.lead_ids ?? []);
    filtered = filtered.filter((lead) => ids.has(lead.id));
  }
  return filtered;
}

function sortDemoLeads(leads: LeadDetail[], params: LeadListParams) {
  const sortBy = params.sort_by ?? "updated_at";
  const dir = params.sort_dir ?? "desc";
  return [...leads].sort((a, b) => {
    const left = a[sortBy as keyof LeadDetail];
    const right = b[sortBy as keyof LeadDetail];
    const comparison = String(left ?? "").localeCompare(String(right ?? ""), undefined, { numeric: true });
    return dir === "asc" ? comparison : -comparison;
  });
}

function unique(values: string[]) {
  return Array.from(new Set(values)).sort();
}
